distributed/comm/tests/test_comms.py

```diff
@@ -289,6 +289,41 @@ def client_communicate(key, delay=0):
     assert set(l) == {1234} | set(range(N))
 
 
+@gen_test()
+def test_comm_failure_threading():
+    """
+    When we fail to connect, make sure we don't make a lot
+    of threads.
+    """
+
+    @gen.coroutine
+    def sleep_for_500ms():
+        max_thread_count = 0
+        for x in range(500):
+            yield gen.sleep(0.001)
+            thread_count = threading.active_count()
+            if thread_count > max_thread_count:
+                max_thread_count = thread_count
+        return max_thread_count
+
+    # tcp.TCPConnector()
+    sleep_future = gen.convert_yielded(sleep_for_500ms())
+    with pytest.raises(IOError):
+        future_connect = gen.convert_yielded(connect("tcp://localhost:28400", 0.6))
+        yield future_connect
+    max_thread_count = yield sleep_future
+    assert max_thread_count == 4
```

Comment:
I recommend testing that no more than a few threads were created since starting this test.  Otherwise this test will fail if some other test leaks a thread (which does happen with a few of our tests).